using_restricted_array.rb

```diff
@@ -8,28 +8,92 @@
 
 ## Calculates the length of the restricted integer array_size
 def length(array)
-  puts "NOT IMPLEMENTED"
+  count = 0
+
+  until array[count] == nil do
+    count += 1
+  end
+
+  return count
+  # puts "NOT IMPLEMENTED"
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  until array[i] == nil do
+    # only print spaces in array that are used
+    if array[i] != SPECIAL_VALUE
+      puts array[i]
+    end
+    i += 1
+  end
+  # puts "NOT IMPLEMENTED"
 end
 
 # Reverses the values in the integer array
 def reverse(array, length) # Ruby
-  puts "NOT IMPLEMENTED"
+  if length <= 1
+    return array
+  end
+
+  i = 0
+  j = length-1
+
+  while i < j do
+    # puts "i: #{i} array[i]: #{array[i]}"
+    # puts "j: #{j} array[j]: #{array[j]}"
+    temp = array[j]
+    array[j] = array[i]
+    array[i] = temp
+    i += 1
+    j -= 1
+  end
+
+  return array
+
+  # puts "NOT IMPLEMENTED"
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  i = 0
+
+  while i < length
+    if array[i] == value_to_find
+      return true
+    else
+      i += 1
+    end
+  end
+  return false
+  # puts "NOT IMPLEMENTED"
 end
 
 # Sorts the array in ascending order.
 def sort(array, length)
-  puts "NOT IMPLEMENTED"
+
+  i = 0
+  j = 1
+  while i < length do
+    # puts "****i: #{i} array#{i}: #{array[i]}"
+    while j < length do
+      # puts "j: #{j} array#{j}: #{array[j]}"
+      if array[j] < array[i]

```

Comment:
Since you've gotten the basics right, just to make you think of further on another optimization: Imagine a case where the input array is sorted in descending order (opposite of what you are looking to do). In this case, you will swap at every step in the inner loop. Could we avoid or reduce this?

Since swapping is three times more work than saving a value, often, a min_index will be tracked. The min_index will be set to be the same as the value of i in the outer loop. In the inner loop, it will get updated to the value of j if array[j] < array[i]. After the inner loop completes, if i and min_index are not the same, only then, do the swap. Let me know if that makes sense.